src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -227,34 +253,174 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+                        }
+
+                        matches.push_back(search.GetFoundLocation());
+                    }
+
+                    // if search box was collapsed or the new one search was triggered - let's cancel this one
+                    if (!_searchState.has_value() || _searchState.value().SearchId != originalSearchId)
+                    {
+                        co_return;
+                    }
+                }
+                _searchState.value().Matches.emplace(matches);
+            }
+        }
+
+        if (auto control{ weakThis.get() })
         {
-            _terminal->SetBlockSelection(false);
-            search.Select();
+            _SelectSearchResult(goForward);
+        }
+    }
+
+    // Method Description:
+    // - Selects one of precomputed search results in the terminal (if exist).
+    // - Updates the search box control accordingly.
+    // - The selection might be preceded by going to next / previous result
+    // - goForward: if true, select next result; if false, select previous result;
+    // if not set, remain at the current result.
+    // Return Value:
+    // - <none>
+    void TermControl::_SelectSearchResult(std::optional<bool> goForward)
+    {
+        if (_searchState.has_value() && _searchState.value().Matches.has_value())
+        {
+            if (goForward.has_value())
+            {
+                _searchState.value().UpdateIndex(goForward.value());
+
+                const auto currentMatch = _searchState.value().GetCurrentMatch();
+                if (currentMatch.has_value())
+                {
+                    auto lock = _terminal->LockForWriting();
+                    _terminal->SetBlockSelection(false);
+                    _terminal->SelectNewRegion(currentMatch.value().first, currentMatch.value().second);
+                    _renderer->TriggerSelection();
+                }
+            }
+
+            if (_searchBox)
+            {
+                _searchBox->SetStatus(_searchState.value().Status());
+                _searchBox->SetNavigationEnabled(!_searchState.value().Matches.value().empty());
+            }
+        }
+    }
+
+    // Method Description:
+    // - Search for a single value in a background
+    // Arguments:
+    // - search: search object to use to find the next match
+    // Return Value:
+    // - <none>
+    winrt::Windows::Foundation::IAsyncOperation<bool> TermControl::_SearchOne(Search& search)
+    {
+        bool found{ false };
+        auto weakThis{ this->get_weak() };
+
+        co_await winrt::resume_background();
+        if (auto control{ weakThis.get() })
+        {
+            // We don't lock the terminal for the duration of the entire search,
+            // since if the terminal was modified the search ID will be updated.
+            auto lock = _terminal->LockForWriting();
+            found = search.FindNext();
+        }
+
+        co_await winrt::resume_foreground(Dispatcher());

```

Comment:
I'm mildly worried that this means for a buffer with 1000 matches, we'll need to hop back and forth between the UI thread and a BG thread 1000 times before we can display the results. 

It _seems_ fine qualitatively, but still...

Reply:
Hmm. We should be able to return the match on whatever queue, and the caller can be in charge of moving to the appropriate queue when it's necessary.